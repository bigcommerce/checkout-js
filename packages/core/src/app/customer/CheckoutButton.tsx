import { CustomerInitializeOptions, CustomerRequestOptions } from '@bigcommerce/checkout-sdk';
import React, { PureComponent } from 'react';

export interface CheckoutButtonProps {
    containerId: string;
    methodId: string;
    deinitialize(options: CustomerRequestOptions): void;
    initialize(options: CustomerInitializeOptions): void;
    onError?(error: Error): void;
}

export default class CheckoutButton extends PureComponent<CheckoutButtonProps> {
    componentDidMount() {
        const { containerId, initialize, methodId, onError } = this.props;

        initialize({
            methodId,
            [methodId]: {
                container: containerId,
                onError,
            },
        });
    }

    componentWillUnmount() {
        const { deinitialize, methodId } = this.props;

        deinitialize({ methodId });
    }

    render() {
        const { containerId } = this.props;

        return <div id={containerId} />;
    }
}
