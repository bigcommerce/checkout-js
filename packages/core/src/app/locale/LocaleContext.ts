import { CurrencyService, LanguageService } from '@bigcommerce/checkout-sdk';
import { createContext } from 'react';

export interface LocaleContextType {
    language: LanguageService;
    date?: {
        inputFormat: string;
    };
    currency?: CurrencyService;
}

const LocaleContext = createContext<LocaleContextType | undefined>(undefined);

export default LocaleContext;
