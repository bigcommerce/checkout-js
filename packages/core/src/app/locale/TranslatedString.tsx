import React, { Fragment, FunctionComponent } from 'react';

import withLanguage, { WithLanguageProps } from './withLanguage';

export interface TranslatedStringProps {
    id: string;
    data?: any;
}

const TranslatedString: FunctionComponent<TranslatedStringProps & WithLanguageProps> = ({
    data,
    id,
    language,
}) => (
    <Fragment>
        { language.translate(id, data) }
    </Fragment>
);

export default withLanguage(TranslatedString);
