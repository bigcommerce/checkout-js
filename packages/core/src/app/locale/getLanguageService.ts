import { createLanguageService, LanguageConfig, LanguageService } from '@bigcommerce/checkout-sdk';

import { FALLBACK_LOCALE, FALLBACK_TRANSLATIONS } from './translations';

let languageService: LanguageService | undefined;

export default function getLanguageService(): LanguageService {
    languageService =
        languageService ??
        createLanguageService({
            fallbackLocale: FALLBACK_LOCALE,
            fallbackTranslations: FALLBACK_TRANSLATIONS,
        });

    return languageService;
}

export type InitializeLanguageService = typeof initializeLanguageService;

export function initializeLanguageService(config: LanguageConfig): LanguageService {
    languageService = createLanguageService({
        ...config,
        defaultLocale: config.locale,
        defaultTranslations: config.defaultTranslations,
        fallbackLocale: FALLBACK_LOCALE,
        fallbackTranslations: FALLBACK_TRANSLATIONS,
    });

    return languageService;
}
