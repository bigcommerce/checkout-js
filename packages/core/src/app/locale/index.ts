export { default as LocaleContext, LocaleContextType } from './LocaleContext';
export { default as createLocaleContext } from './createLocaleContext';
export { default as getDefaultTranslations } from './getDefaultTranslations';
export { default as getLanguageService, InitializeLanguageService } from './getLanguageService';
export { default as masterpassFormatLocale } from './masterpassFormatLocale';
export { default as withCurrency, WithCurrencyProps } from './withCurrency';
export { default as withLanguage, WithLanguageProps } from './withLanguage';
export { default as withDate, WithDateProps } from './withDate';
export { default as LanguageWindow, isLanguageWindow } from './LanguageWindow';
export { default as LocaleProvider } from './LocaleProvider';
export { default as TranslatedHtml, TranslatedHtmlProps } from './TranslatedHtml';
export { default as TranslatedLink, TranslatedLinkProps } from './TranslatedLink';
export { default as TranslatedString, TranslatedStringProps } from './TranslatedString';
