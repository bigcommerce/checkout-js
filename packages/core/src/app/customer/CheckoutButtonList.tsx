import { CustomerInitializeOptions, CustomerRequestOptions } from '@bigcommerce/checkout-sdk';
import React, { memo, Fragment, FunctionComponent } from 'react';

import { isApplePayWindow } from '../common/utility';
import { TranslatedString } from '../locale';

import { ApplePayButton } from './customWalletButton';
import CheckoutButton from './CheckoutButton';

const APPLE_PAY = 'applepay';

// TODO: The API should tell UI which payment method offers its own checkout button
export const SUPPORTED_METHODS: string[] = [
    'amazon',
    'amazonpay',
    APPLE_PAY,
    'braintreevisacheckout',
    'chasepay',
    'masterpass',
    'googlepayadyenv2',
    'googlepayadyenv3',
    'googlepayauthorizenet',
    'googlepaybraintree',
    'googlepaycheckoutcom',
    'googlepaycybersourcev2',
    'googlepayorbital',
    'googlepaystripe',
    'googlepaystripeupe',
];

export interface CheckoutButtonListProps {
    methodIds?: string[];
    isInitializing?: boolean;
    checkEmbeddedSupport?(methodIds: string[]): void;
    deinitialize(options: CustomerRequestOptions): void;
    initialize(options: CustomerInitializeOptions): void;
    onError?(error: Error): void;
}

const CheckoutButtonList: FunctionComponent<CheckoutButtonListProps> = ({
    checkEmbeddedSupport,
    onError,
    isInitializing = false,
    methodIds,
    ...rest
}) => {
    const supportedMethodIds = (methodIds ?? []).filter(methodId => {
        if (methodId === APPLE_PAY && !isApplePayWindow(window)) {
            return false;
        }

        return SUPPORTED_METHODS.indexOf(methodId) !== -1;
    });

    if (supportedMethodIds.length === 0) {
        return null;
    }

    if (checkEmbeddedSupport) {
        try {
            checkEmbeddedSupport(supportedMethodIds);
        } catch (error) {
            if (onError) {
                onError(error);
            } else {
                throw error;
            }

            return null;
        }
    }

    return (
        <Fragment>
            { !isInitializing && <p><TranslatedString id="remote.continue_with_text" /></p> }

            <div className="checkoutRemote">
                { supportedMethodIds.map(methodId =>
                    methodId === 'applepay' ?
                        <ApplePayButton
                            containerId={ `${methodId}CheckoutButton` }
                            key={ methodId }
                            methodId={ methodId }
                            onError={ onError }
                            { ...rest }
                        /> :
                        <CheckoutButton
                            containerId={ `${methodId}CheckoutButton` }
                            key={ methodId }
                            methodId={ methodId }
                            onError={ onError }
                            { ...rest }
                        />
                ) }
            </div>
        </Fragment>
    );
};

export default memo(CheckoutButtonList);
