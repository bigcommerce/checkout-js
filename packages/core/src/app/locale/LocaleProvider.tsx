import { CheckoutService, createCurrencyService, StoreConfig } from '@bigcommerce/checkout-sdk';
import { memoizeOne } from '@bigcommerce/memoize';
import React, { Component, ReactNode } from 'react';

import getLanguageService from './getLanguageService';
import LocaleContext from './LocaleContext';

export interface LocaleProviderProps {
    checkoutService: CheckoutService;
}

export interface LocaleProviderState {
    config?: StoreConfig;
}

class LocaleProvider extends Component<LocaleProviderProps> {
    state: Readonly<LocaleProviderState> = {};

    private languageService = getLanguageService();
    private unsubscribe?: () => void;

    private getContextValue = memoizeOne((config?: StoreConfig) => {
        return {
            currency: config ? createCurrencyService(config) : undefined,
            date: config
                ? {
                      inputFormat: config.inputDateFormat,
                  }
                : undefined,
            language: this.languageService,
        };
    });

    componentDidMount(): void {
        const { checkoutService } = this.props;

        this.unsubscribe = checkoutService.subscribe(
            ({ data }) => {
                this.setState({ config: data.getConfig() });
            },
            ({ data }) => data.getConfig(),
        );
    }

    componentWillUnmount(): void {
        if (this.unsubscribe) {
            this.unsubscribe();
            this.unsubscribe = undefined;
        }
    }

    render(): ReactNode {
        const { children } = this.props;
        const { config } = this.state;

        return (
            <LocaleContext.Provider value={this.getContextValue(config)}>
                {children}
            </LocaleContext.Provider>
        );
    }
}

export default LocaleProvider;
