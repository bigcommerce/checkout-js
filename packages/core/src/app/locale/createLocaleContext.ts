import { createCurrencyService, StoreConfig } from '@bigcommerce/checkout-sdk';

import getLanguageService from './getLanguageService';
import { LocaleContextType } from './LocaleContext';

export default function createLocaleContext(config: StoreConfig): Required<LocaleContextType> {
    if (!config) {
        throw new Error('Missing configuration data');
    }

    const { inputDateFormat } = config;

    return {
        currency: createCurrencyService(config),
        date: {
            inputFormat: inputDateFormat,
        },
        language: getLanguageService(),
    };
}
