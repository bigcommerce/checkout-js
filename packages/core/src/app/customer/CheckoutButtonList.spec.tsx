import { mount, render } from 'enzyme';
import { noop } from 'lodash';
import React from 'react';

import { getStoreConfig } from '../config/config.mock';
import { createLocaleContext, LocaleContext, LocaleContextType, TranslatedString } from '../locale';

import CheckoutButton from './CheckoutButton';
import CheckoutButtonList from './CheckoutButtonList';

describe('CheckoutButtonList', () => {
    let localeContext: LocaleContextType;

    beforeEach(() => {
        localeContext = createLocaleContext(getStoreConfig());
    });

    it('matches snapshot', () => {
        const component = render(
            <LocaleContext.Provider value={localeContext}>
                <CheckoutButtonList
                    deinitialize={noop}
                    initialize={noop}
                    methodIds={['amazon', 'braintreevisacheckout']}
                />
            </LocaleContext.Provider>,
        );

        expect(component).toMatchSnapshot();
    });

    it('filters out unsupported methods', () => {
        const component = mount(
            <LocaleContext.Provider value={localeContext}>
                <CheckoutButtonList
                    deinitialize={noop}
                    initialize={noop}
                    methodIds={['amazon', 'braintreevisacheckout', 'foobar']}
                />
            </LocaleContext.Provider>,
        );

        expect(component.find(CheckoutButton)).toHaveLength(2);
    });

    it('does not crash when no methods are passed', () => {
        const component = mount(
            <LocaleContext.Provider value={localeContext}>
                <CheckoutButtonList deinitialize={noop} initialize={noop} />
            </LocaleContext.Provider>,
        );

        expect(component.html()).toBeFalsy();
    });

    it('does not render if there are no supported methods', () => {
        const component = mount(
            <LocaleContext.Provider value={localeContext}>
                <CheckoutButtonList deinitialize={noop} initialize={noop} methodIds={['foobar']} />
            </LocaleContext.Provider>,
        );

        expect(component.html()).toBeFalsy();
    });

    it('does not render the translated string when initializing', () => {
        const component = mount(
            <LocaleContext.Provider value={localeContext}>
                <CheckoutButtonList
                    deinitialize={noop}
                    initialize={noop}
                    isInitializing={true}
                    methodIds={['amazon', 'braintreevisacheckout']}
                />
            </LocaleContext.Provider>,
        );

        expect(component.find(TranslatedString)).toHaveLength(0);
    });

    it('passes data to every checkout button', () => {
        const deinitialize = jest.fn();
        const initialize = jest.fn();
        const component = mount(
            <LocaleContext.Provider value={localeContext}>
                <CheckoutButtonList
                    deinitialize={deinitialize}
                    initialize={initialize}
                    methodIds={['amazon', 'braintreevisacheckout']}
                />
            </LocaleContext.Provider>,
        );

        expect(component.find(CheckoutButton).at(0).props()).toEqual({
            containerId: 'amazonCheckoutButton',
            methodId: 'amazon',
            deinitialize,
            initialize,
        });
    });

    it('notifies parent if methods are incompatible with Embedded Checkout', () => {
        const methodIds = ['amazon', 'braintreevisacheckout'];
        const onError = jest.fn();
        const checkEmbeddedSupport = jest.fn(() => {
            throw new Error();
        });

        render(
            <LocaleContext.Provider value={localeContext}>
                <CheckoutButtonList
                    checkEmbeddedSupport={checkEmbeddedSupport}
                    deinitialize={noop}
                    initialize={noop}
                    methodIds={methodIds}
                    onError={onError}
                />
            </LocaleContext.Provider>,
        );

        expect(checkEmbeddedSupport).toHaveBeenCalledWith(methodIds);

        expect(onError).toHaveBeenCalledWith(expect.any(Error));
    });
});
