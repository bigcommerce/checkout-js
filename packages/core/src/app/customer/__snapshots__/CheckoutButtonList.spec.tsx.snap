// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutButtonList matches snapshot 1`] = `
Array [
  <p>
    Or continue with
  </p>,
  <div
    class="checkoutRemote"
  >
    <div
      id="amazonCheckoutButton"
    />
    <div
      id="braintreevisacheckoutCheckoutButton"
    />
  </div>,
]
`;
